File: backend/app/schemas.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class SheetInfo(BaseModel):
    name: str
    rows: int = Field(ge=0)
    columns: int = Field(ge=0)
    has_data: bool


class WorkbookInfo(BaseModel):
    file_id: str
    original_name: str
    safe_name: str
    size_bytes: int
    mime_type: str
    sheets: list[SheetInfo]
    created_at: datetime


class ColumnProfile(BaseModel):
    name: str
    inferred_type: Literal["date", "number", "category", "text", "unknown"]
    semantic_role: Literal["date", "dimension", "measure", "identifier", "unknown"]
    null_count: int
    unique_count: int
    sample_values: list[Any] = Field(default_factory=list)
    ambiguous: bool = False
    reason: str = ""


class QualityReport(BaseModel):
    row_count: int
    column_count: int
    missing_cells: int
    missing_rate: float = Field(ge=0, le=1)
    duplicate_rows: int
    invalid_values: int
    outlier_count: int
    formula_like_cells: int
    score: int = Field(ge=0, le=100)
    notes: list[str] = Field(default_factory=list)


class AnalysisPlanContent(BaseModel):
    objective: str = Field(min_length=10, max_length=500)
    measures: list[str] = Field(default_factory=list)
    dimensions: list[str] = Field(default_factory=list)
    dates: list[str] = Field(default_factory=list)
    chart_strategy: list[Literal["trend", "category_comparison", "share", "distribution"]] = Field(min_length=1)
    privacy: str = Field(min_length=10, max_length=500)


class AnalysisPlan(AnalysisPlanContent):
    mode: Literal["mock", "ollama", "groq"]
    model: str


class ResultValue(BaseModel):
    value: int | float | str
    operation: str
    source_columns: list[str]
    query: str


class KpiSpec(BaseModel):
    id: str
    label: str
    result_ref: str
    format: Literal["number", "currency", "percent", "decimal"]
    tone: Literal["primary", "positive", "neutral", "warning"] = "neutral"


class ChartSeries(BaseModel):
    name: str
    values: list[float]


class ChartSpec(BaseModel):
    id: str
    title: str
    type: Literal["line", "bar", "stacked_bar", "area", "donut", "scatter", "histogram", "heatmap", "boxplot"]
    categories: list[str] = Field(default_factory=list)
    series: list[ChartSeries] = Field(default_factory=list)
    result_refs: list[str] = Field(min_length=1)
    x_label: str = ""
    y_label: str = ""


class TableSpec(BaseModel):
    id: str
    title: str
    columns: list[str]
    rows: list[dict[str, Any]]


class FilterSpec(BaseModel):
    column: str
    label: str
    values: list[str]


class InsightSpec(BaseModel):
    title: str
    text: str
    result_refs: list[str] = Field(default_factory=list)


class DashboardSpec(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    title: str = Field(min_length=3)
    description: str = Field(min_length=10)
    kpis: list[KpiSpec] = Field(default_factory=list)
    charts: list[ChartSpec] = Field(min_length=1)
    tables: list[TableSpec] = Field(min_length=1)
    filters: list[FilterSpec] = Field(default_factory=list)
    computed_results: dict[str, ResultValue] = Field(min_length=1)
    value_formats: dict[str, str] = Field(default_factory=dict)
    layout: list[str] = Field(min_length=1)
    warnings: list[str] = Field(default_factory=list)
    quality_notes: list[str] = Field(default_factory=list)
    executive_summary: str = Field(min_length=10)
    detailed_insights: list[InsightSpec] = Field(default_factory=list)
    dimensions: list[str] = Field(default_factory=list)
    measures: list[str] = Field(default_factory=list)

    @model_validator(mode="after")
    def all_numeric_claims_are_referenced(self) -> "DashboardSpec":
        known = set(self.computed_results)
        refs = [k.result_ref for k in self.kpis]
        refs += [ref for chart in self.charts for ref in chart.result_refs]
        refs += [ref for insight in self.detailed_insights for ref in insight.result_refs]
        missing = sorted(set(refs) - known)
        if missing:
            raise ValueError(f"مراجع نتائج غير موثقة: {', '.join(missing)}")
        return self


class MissingValueOverride(BaseModel):
    column: str = Field(min_length=1)
    source_row: int = Field(ge=2)
    value: str | int | float


class AnalysisStart(BaseModel):
    file_id: str
    sheet_name: str
    max_iterations: int = Field(default=3, ge=1, le=5)
    column_mapping: dict[str, str] = Field(default_factory=dict)
    missing_value_mode: Literal["recommended", "manual"] = "recommended"
    missing_value_overrides: list[MissingValueOverride] = Field(default_factory=list)


class ClarificationAnswer(BaseModel):
    mappings: dict[str, str] = Field(min_length=1)


class AnalysisQuestion(BaseModel):
    question: str = Field(min_length=2, max_length=500)


class AnalysisAnswer(BaseModel):
    answer: str = Field(min_length=2)
    sources: list[str] = Field(default_factory=list)


class AgentExecution(BaseModel):
    agent: Literal["cleaning_agent", "analysis_agent", "dashboard_agent"]
    label: str
    responsibility: str
    status: Literal["completed", "failed"]
    summary: str
    artifacts: list[str] = Field(default_factory=list)


class CleaningImputation(BaseModel):
    column: str = Field(min_length=1)
    count: int = Field(ge=1)
    strategy: Literal["derived", "sequential", "mean", "median", "label", "manual", "retained"]
    fill_value: float | str | None = None
    source_rows: list[int] = Field(default_factory=list)
    explanation: str = Field(min_length=10)


class CleaningAudit(BaseModel):
    input_rows: int = Field(ge=0)
    output_rows: int = Field(ge=0)
    excluded_summary_rows: list[int] = Field(default_factory=list)
    numeric_conversions: int = Field(ge=0)
    date_conversions: int = Field(ge=0)
    normalized_text_cells: int = Field(ge=0)
    invalid_numeric_cells: int = Field(ge=0)
    invalid_date_cells: int = Field(ge=0)
    excluded_empty_columns: list[str] = Field(default_factory=list)
    formula_calculations: int = Field(default=0, ge=0)
    missing_value_mode: Literal["recommended", "manual"] = "recommended"
    missing_values_before: dict[str, int] = Field(default_factory=dict)
    missing_locations: dict[str, list[int]] = Field(default_factory=dict)
    output_source_rows: list[int] = Field(default_factory=list)
    remaining_missing_values: dict[str, int] = Field(default_factory=dict)
    imputation_actions: list[CleaningImputation] = Field(default_factory=list)
    removed_duplicate_rows: list[int] = Field(default_factory=list)
    policy: str = Field(min_length=10)

    @model_validator(mode="after")
    def output_cannot_exceed_input(self) -> "CleaningAudit":
        if self.output_rows > self.input_rows:
            raise ValueError("عدد صفوف نسخة التحليل لا يمكن أن يتجاوز صفوف المصدر أثناء التنظيف.")
        return self


class CustomCalculationRequest(BaseModel):
    instruction: str = Field(min_length=5, max_length=500)


class CustomCalculationResponse(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    expression: str
    value: float
    format: Literal["percent", "decimal"]
    source_columns: list[str] = Field(min_length=1)
    verification: str
    query: str


class AnalysisResponse(BaseModel):
    analysis_id: str
    status: Literal["queued", "running", "waiting_for_clarification", "completed", "completed_with_fallback", "failed"]
    stage: str
    progress: int = Field(ge=0, le=100)
    ambiguity: dict[str, Any] | None = None
    analysis_plan: AnalysisPlan | None = None
    dashboard: DashboardSpec | None = None
    quality: QualityReport | None = None
    cleaning_audit: CleaningAudit | None = None
    agent_runs: list[AgentExecution] = Field(default_factory=list)
    trace: list[str] = Field(default_factory=list)
    error: str | None = None


class PreviewResponse(BaseModel):
    file_id: str
    sheet_name: str
    columns: list[ColumnProfile]
    rows: list[dict[str, Any]]
    total_rows: int
    cleaning_audit: CleaningAudit | None = None


class HealthResponse(BaseModel):
    status: Literal["ok", "degraded"] = "ok"
    service: Literal["bayyinah-backend"] = "bayyinah-backend"
    mode: Literal["mock", "ollama", "groq"]
    model: str
    llm_ready: bool
    detail: str
    database: Literal["sqlite", "postgresql"]
    jobs: Literal["inline", "background", "celery"]

File: frontend/lib/schemas.ts
import { z } from "zod";

export const sheetSchema = z.object({ name: z.string(), rows: z.number(), columns: z.number(), has_data: z.boolean() });
export const healthSchema = z.object({
  status: z.enum(["ok", "degraded"]), service: z.literal("bayyinah-backend"),
  mode: z.enum(["mock", "ollama", "groq"]), model: z.string(), llm_ready: z.boolean(), detail: z.string(),
  database: z.enum(["sqlite", "postgresql"]), jobs: z.enum(["inline", "background", "celery"]),
});
export const workbookSchema = z.object({
  file_id: z.string(), original_name: z.string(), safe_name: z.string(), size_bytes: z.number(),
  mime_type: z.string(), sheets: z.array(sheetSchema), created_at: z.string(),
});
const resultValueSchema = z.object({ value: z.union([z.string(), z.number()]), operation: z.string(), source_columns: z.array(z.string()), query: z.string() });
const chartSchema = z.object({
  id: z.string(), title: z.string(), type: z.string(), categories: z.array(z.string()),
  series: z.array(z.object({ name: z.string(), values: z.array(z.number()) })), result_refs: z.array(z.string()),
  x_label: z.string(), y_label: z.string(),
});
export const dashboardSchema = z.object({
  title: z.string(), description: z.string(),
  kpis: z.array(z.object({ id: z.string(), label: z.string(), result_ref: z.string(), format: z.string(), tone: z.string() })),
  charts: z.array(chartSchema),
  tables: z.array(z.object({ id: z.string(), title: z.string(), columns: z.array(z.string()), rows: z.array(z.record(z.string(), z.unknown())) })),
  filters: z.array(z.object({ column: z.string(), label: z.string(), values: z.array(z.string()) })),
  computed_results: z.record(z.string(), resultValueSchema), value_formats: z.record(z.string(), z.string()),
  layout: z.array(z.string()), warnings: z.array(z.string()), quality_notes: z.array(z.string()),
  executive_summary: z.string(), detailed_insights: z.array(z.object({ title: z.string(), text: z.string(), result_refs: z.array(z.string()) })),
  dimensions: z.array(z.string()).default([]), measures: z.array(z.string()).default([]),
});
const cleaningAuditSchema = z.object({
  input_rows: z.number(), output_rows: z.number(), excluded_summary_rows: z.array(z.number()),
  numeric_conversions: z.number(), date_conversions: z.number(), normalized_text_cells: z.number(),
  invalid_numeric_cells: z.number(), invalid_date_cells: z.number(),
  excluded_empty_columns: z.array(z.string()).default([]), formula_calculations: z.number().default(0),
  missing_value_mode: z.enum(["recommended", "manual"]).default("recommended"),
  missing_values_before: z.record(z.string(), z.number()).default({}),
  missing_locations: z.record(z.string(), z.array(z.number())).default({}),
  output_source_rows: z.array(z.number()).default([]),
  remaining_missing_values: z.record(z.string(), z.number()).default({}),
  imputation_actions: z.array(z.object({
    column: z.string(), count: z.number(), strategy: z.enum(["derived", "sequential", "mean", "median", "label", "manual", "retained"]),
    fill_value: z.union([z.string(), z.number()]).nullable(), source_rows: z.array(z.number()).default([]), explanation: z.string(),
  })).default([]), removed_duplicate_rows: z.array(z.number()).default([]),
  policy: z.string(),
});
export const previewSchema = z.object({
  file_id: z.string(), sheet_name: z.string(), total_rows: z.number(),
  columns: z.array(z.object({ name: z.string(), inferred_type: z.string(), semantic_role: z.string(), null_count: z.number(), unique_count: z.number(), sample_values: z.array(z.unknown()), ambiguous: z.boolean(), reason: z.string() })),
  rows: z.array(z.record(z.string(), z.unknown())), cleaning_audit: cleaningAuditSchema.nullable().default(null),
});
export const analysisSchema = z.object({
  analysis_id: z.string(), status: z.enum(["queued", "running", "waiting_for_clarification", "completed", "completed_with_fallback", "failed"]),
  stage: z.string(), progress: z.number(), ambiguity: z.record(z.string(), z.unknown()).nullable(),
  analysis_plan: z.object({
    mode: z.enum(["mock", "ollama", "groq"]), model: z.string(), objective: z.string(),
    measures: z.array(z.string()), dimensions: z.array(z.string()), dates: z.array(z.string()),
    chart_strategy: z.array(z.enum(["trend", "category_comparison", "share", "distribution"])),
    privacy: z.string(),
  }).nullable(),
  dashboard: dashboardSchema.nullable(), quality: z.object({
    row_count: z.number(), column_count: z.number(), missing_cells: z.number(), missing_rate: z.number(),
    duplicate_rows: z.number(), invalid_values: z.number(), outlier_count: z.number(), formula_like_cells: z.number(),
    score: z.number(), notes: z.array(z.string()),
  }).nullable(), cleaning_audit: cleaningAuditSchema.nullable().default(null), trace: z.array(z.string()), error: z.string().nullable(),
  agent_runs: z.array(z.object({
    // Keep the client compatible with older deployments that used aliases.
    // The UI normalizes the value to one of the three canonical agent ids.
    agent: z.string(),
    label: z.string(), responsibility: z.string(), status: z.enum(["completed", "failed"]),
    summary: z.string(), artifacts: z.array(z.string()),
  })).default([]),
});
export const analysisAnswerSchema = z.object({ answer: z.string(), sources: z.array(z.string()) });
export const customCalculationSchema = z.object({
  name: z.string(), expression: z.string(), value: z.number(), format: z.enum(["percent", "decimal"]),
  source_columns: z.array(z.string()), verification: z.string(), query: z.string(),
});

export type Workbook = z.infer<typeof workbookSchema>;
export type Health = z.infer<typeof healthSchema>;
export type Preview = z.infer<typeof previewSchema>;
export type Analysis = z.infer<typeof analysisSchema>;
export type Dashboard = z.infer<typeof dashboardSchema>;
export type AnalysisAnswer = z.infer<typeof analysisAnswerSchema>;
export type CustomCalculation = z.infer<typeof customCalculationSchema>;